src/Features/Core/Portable/ConvertLinq/AbstractConvertForEachToLinqQueryProvider.cs

```diff
@@ -0,0 +1,145 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.ConvertLinq
+{
+    internal abstract class AbstractConvertForEachToLinqQueryProvider<TForEachStatement, TStatement> : CodeRefactoringProvider
+        where TForEachStatement : TStatement
+        where TStatement : SyntaxNode
+    {
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var forEachStatement = FindNodeToRefactor(root, context.Span);
+            if (forEachStatement == null)
+            {
+                return;
+            }
+
+            if (!Validate(forEachStatement))
+            {
+                return;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            if (TryBuildConverter(forEachStatement, semanticModel, cancellationToken, out IConverter converter) &&
+                !semanticModel.GetDiagnostics(forEachStatement.Span, cancellationToken).Any(diagnostic => diagnostic.DefaultSeverity == DiagnosticSeverity.Error))
+            {
+                context.RegisterRefactoring(new ForEachToLinqQueryCodeAction(Title, c =>
+                {
+                    var editor = new SyntaxEditor(semanticModel.SyntaxTree.GetRoot(c), document.Project.Solution.Workspace);
+                    converter.Convert(editor, semanticModel, c);
+                    var newRoot = editor.GetChangedRoot();
+                    var rootWithLinqUsing = AddLinqUsing(newRoot);
+                    return Task.FromResult(document.WithSyntaxRoot(rootWithLinqUsing));
```

Comment:
can you extract all of this into a helper method.  that way this function doesn't accidently capture anything it should not?